Allowing WORLD and LOCAL_WORLD_ALIGNED in ContactModel1D

skleff1994 opened the pull request:
Minor enhancements for the 1D contact model . Related to #812 

- Add a mask for constraint dimension selection . The user selects optionally `type` in `{0,1,2}` for `{x,y,z}` the z-axis being the default one
- Add optional argument of pinocchio reference frame in {`LOCAL`, `WORLD` , `LOCAL_WORLD_ALIGNED`}, `LOCAL` being the default one

The mask could be extended to 2D contact , and the pinocchio reference frame option could be propagated to other contact models , if deemed useful . But I think all of this will be handled in Pinocchio in the future...

jcarpent commented on bindings/python/crocoddyl/multibody/contacts/contact-1d.cpp:
```diff
@@ -15,6 +15,12 @@ namespace python {
 void exposeContact1D() {
   bp::register_ptr_to_python<boost::shared_ptr<ContactModel1D> >();
 
+  bp::enum_<pinocchio::ReferenceFrame>("ReferenceFrame")
```
You should not duplicate this. This is a Pinocchio feature.

jcarpent replied:
If you want to have this thing exposed in Crocoddyl, there is a specific helper function to import ReferenceFrame within the Croco namespace, assuming Pinocchio is exposed first.

cmastalli replied:
Let's do not expose this in Crocoddyl.
There is no reason to duplicate this code.

skleff1994 replied:
[fd58261](https://github.com/loco-3d/crocoddyl/pull/1054/commits/fd582614a34b626d8cb0751df8132d31f643845b)

jcarpent commented on bindings/python/crocoddyl/multibody/contacts/contact-1d.cpp:
```diff
@@ -24,21 +30,23 @@ void exposeContact1D() {
-               bp::optional<Eigen::Vector2d> >(
-          bp::args("self", "state", "id", "xref", "nu", "gains"),
+               bp::optional<Eigen::Vector2d, std::size_t, pinocchio::ReferenceFrame> >(
+          bp::args("self", "state", "id", "xref", "nu", "gains", "type", "pinRefFrame"),
```
Set default argument here.

cmastalli replied:
Also rename the arguments as follows:
 - `type` to `mask`
 - `pinRefFrame` to `type`

This latter follows the standard used in other classes.

skleff1994 replied:
@jcarpent  concerning the default argument I had

`( bp::args("self", "state", "id", "xref", "nu", "gains", "type"), bp::arg("pinRefFrame")=pinocchio::LOCAL ),`

but it seemed to make python unittest fail and caused an import error in python, I couldn't figure out why . I need to investigate this . Any clue?

skleff1994 replied:
@cmastalli  addressed in [fd58261](https://github.com/loco-3d/crocoddyl/pull/1054/commits/fd582614a34b626d8cb0751df8132d31f643845b)

jcarpent commented on include/crocoddyl/multibody/contacts/contact-1d.hxx:
```diff
@@ -9,17 +9,36 @@
+                                             const Scalar xref, 
+                                             const std::size_t nu, 
+                                             const Vector2s& gains, 
+                                             const std::size_t& type, 
```
The reference is not needed as it is a basic type.

cmastalli replied:
Better, please define an enum for `type`.

skleff1994 replied:
addressed in [fd58261](https://github.com/loco-3d/crocoddyl/pull/1054/commits/fd582614a34b626d8cb0751df8132d31f643845b)

cmastalli commented on bindings/python/crocoddyl/multibody/contacts/contact-1d.cpp:
```diff
@@ -24,21 +30,23 @@ void exposeContact1D() {
-          ":param gains: gains of the contact model (default np.matrix([0.,0.]))"))
-      .def(bp::init<boost::shared_ptr<StateMultibody>, pinocchio::FrameIndex, double, bp::optional<Eigen::Vector2d> >(
+          ":param gains: gains of the contact model (default np.matrix([0.,0.]))\n"
+          ":param type: axis of the contact constraint (0=x, 1=y or 2=z)\n"
+          ":param pinRefFrame: pin.ReferenceFrame in {LOCAL, WORLD, LOCAL_WORLD_ALIGNED}"))
```
Please adapt the documentation as suggested above

skleff1994 replied:
addressed in [fd58261](https://github.com/loco-3d/crocoddyl/pull/1054/commits/fd582614a34b626d8cb0751df8132d31f643845b)

cmastalli commented on bindings/python/crocoddyl/multibody/contacts/contact-1d.cpp:
```diff
@@ -24,21 +30,23 @@ void exposeContact1D() {
           ":param id: reference frame id of the contact\n"
           ":param xref: contact position used for the Baumgarte stabilization\n"
-          ":param gains: gains of the contact model (default np.matrix([0.,0.]))"))
+          ":param gains: gains of the contact model (default np.matrix([0.,0.]))\n"))
```
Revert this back

skleff1994 replied:
addressed in [fd58261](https://github.com/loco-3d/crocoddyl/pull/1054/commits/fd582614a34b626d8cb0751df8132d31f643845b)

cmastalli commented on bindings/python/crocoddyl/multibody/contacts/contact-1d.cpp:
```diff
@@ -68,7 +76,10 @@ void exposeContact1D() {
                     bp::make_function(&ContactModel1D::get_gains, bp::return_value_policy<bp::return_by_value>()),
-                    "contact gains");
+                    "contact gains")
+      .add_property("pinReferenceFrame",
```
1. Renamed this as suggested above
2. Add the getter and setter for the `mask`

skleff1994 replied:
addressed in [fd58261](https://github.com/loco-3d/crocoddyl/pull/1054/commits/fd582614a34b626d8cb0751df8132d31f643845b)

cmastalli commented on bindings/python/crocoddyl/multibody/contacts/contact-1d.cpp:
```diff
@@ -68,7 +76,10 @@ void exposeContact1D() {
+                    "contact gains")
+      .add_property("pinReferenceFrame",
+                    bp::make_function(&ContactModel1D::get_pinReferenceFrame, bp::return_value_policy<bp::return_by_value>()),
+                    &ContactModel1D::set_pinReferenceFrame, "pin.ReferenceFrame");
```
Change documentation to `reference type of contact`

skleff1994 replied:
addressed in [fd58261](https://github.com/loco-3d/crocoddyl/pull/1054/commits/fd582614a34b626d8cb0751df8132d31f643845b)

cmastalli commented on include/crocoddyl/multibody/contacts/contact-1d.hpp:
```diff
@@ -37,6 +37,7 @@ class ContactModel1DTpl : public ContactModelAbstractTpl<_Scalar> {
   typedef typename MathBase::Vector3s Vector3s;
   typedef typename MathBase::VectorXs VectorXs;
   typedef typename MathBase::Matrix3s Matrix3s;
+  typedef typename MathBase::Vector6s Vector6s;
```
Remove this line. It doesn't seem to be needed

skleff1994 replied:
addressed in [fd58261](https://github.com/loco-3d/crocoddyl/pull/1054/commits/fd582614a34b626d8cb0751df8132d31f643845b)

cmastalli commented on include/crocoddyl/multibody/contacts/contact-1d.hxx:
```diff
@@ -9,17 +9,36 @@
+                                             const pinocchio::FrameIndex id,
+                                             const Scalar xref, 
+                                             const Vector2s& gains,
+                                             const pinocchio::ReferenceFrame pinRefFrame)
```
Change the name of the arguments as suggested above.

skleff1994 replied:
addressed in [fd58261](https://github.com/loco-3d/crocoddyl/pull/1054/commits/fd582614a34b626d8cb0751df8132d31f643845b)

cmastalli commented on include/crocoddyl/multibody/contacts/contact-1d.hxx:
```diff
@@ -9,17 +9,36 @@
+                                             const Vector2s& gains, 
+                                             const std::size_t& type, 
+                                             const pinocchio::ReferenceFrame pinRefFrame)
+    : Base(state, 1, nu), xref_(xref), gains_(gains), type_(type), pinReferenceFrame_(pinRefFrame) {
```
Change the name of the arguments as suggested above

skleff1994 replied:
addressed in [fd58261](https://github.com/loco-3d/crocoddyl/pull/1054/commits/fd582614a34b626d8cb0751df8132d31f643845b)

cmastalli commented on include/crocoddyl/multibody/contacts/contact-1d.hxx:
```diff
@@ -9,17 +9,36 @@
+                                             const pinocchio::ReferenceFrame pinRefFrame)
+    : Base(state, 1, nu), xref_(xref), gains_(gains), type_(type), pinReferenceFrame_(pinRefFrame) {
   id_ = id;
+  if(type_ < 0 || type_ > 2){
```
This doesn't follow the code format. Please run the code formatter as described in the wiki

skleff1994 replied:
addressed in [fd58261](https://github.com/loco-3d/crocoddyl/pull/1054/commits/fd582614a34b626d8cb0751df8132d31f643845b)

cmastalli reviewed (changes requested):
There are still many changes to be tackled.
However, I don't understand why you need to define a reference frame that is not LOCAL.
These classes are not aimed to track a reference acceleration for instance.
So, why?

skleff1994 commented:
Thanks for your review , I will now implement some of the requested changes.
 
I also noticed that the [contact force residual computation](https://github.com/skleff1994/crocoddyl/blob/31acf4adb4e556e15e34eaabe2196a447f6c082e/include/crocoddyl/multibody/residuals/contact-force.hxx#L44) needs to be updated as well  , and therefore be aware of the mask of the contact model .

@cmastalli  the purpose of this enhancement is to allow force computations in the WORLD frame, which is necessary for e.g. sanding tasks where the desired force is defined in WORLD (the LOCAL z-axis is not necessarily aligned with the vertical direction).  In that case , I believe we need to perform all contact-related computations in LOCAL_WORLD_ALIGNED , which is currently not allowed by Crocoddyl.

@jcarpent I believe that Baumgarte stabilizing term is already expressed in WORLD for all contact models - which I also don't really understand as the acceleration and force computations are currently in LOCAL .  Wouldn't it make sense to compute the Baumgarte terms according in the pinocchio reference frame ?

skleff1994 commented:
Concerning the force residual, I was thinking I could add the mask to ContactData1D so that it can be accessed in the force residual computation ( we need to select x, y, or z component depending on the mask of the contact model ). Also , we should pass the pinocchio reference as an optional argument in force residual constructor to be able to compute the 1D force residual in any frame . From there, this could be straightforwardly extended to other contact models I believe . What do you think ?

skleff1994 commented on include/crocoddyl/multibody/contacts/contact-1d.hxx:
```diff
@@ -30,22 +49,25 @@ void ContactModel1DTpl<Scalar>::calc(const boost::shared_ptr<ContactDataAbstract
 
   if (gains_[0] != 0.) {
-    d->a0[0] += gains_[0] * (d->pinocchio->oMf[id_].translation()[2] - xref_);
+    if(pinReferenceFrame_ == pinocchio::WORLD){
```
I believe that Baumgarte stabilizing term is already expressed in WORLD for all contact models - which I also don't really understand as the acceleration and force computations are currently in LOCAL .  Wouldn't it make sense to compute the Baumgarte terms in the pinocchio reference frame ?

cmastalli commented on include/crocoddyl/multibody/contacts/contact-1d.hxx:
```diff
@@ -90,9 +100,23 @@ void ContactModel1DTpl<Scalar>::updateForce(const boost::shared_ptr<ContactDataA
+      break;
+    }
+    case pinocchio::LOCAL_WORLD_ALIGNED: {
+      jMc = pinocchio::SE3(Matrix3s::Identity(), d->jMf.translation());
```
This is allocating data dynamically.
Instead, you should modify the translation only, i.e.
```suggestion
 jMc_.translation = d->jMf.translation();
```

Then, you have to set the identity matrix only when you have define `LOCAL_WORLD_ALIGNED`

cmastalli replied:
This is allocating data dynamically.
Instead, you should modify the translation only, i.e.
```suggestion
 jMc_.translation = d->jMf.translation();
```

Then, you have to set the identity matrix only when you have define `LOCAL_WORLD_ALIGNED`

cmastalli commented on include/crocoddyl/multibody/contacts/contact-1d.hxx:
```diff
@@ -90,9 +100,23 @@ void ContactModel1DTpl<Scalar>::updateForce(const boost::shared_ptr<ContactDataA
-  const Eigen::Ref<const Matrix3s> R = d->jMf.rotation();
-  data->f.linear() = R.col(2) * force[0];
-  data->f.angular() = d->jMf.translation().cross(data->f.linear());
+  pinocchio::SE3 jMc;
```
Let's define a global variable for this, and called it `jMc_`.
This avoids to allocate data dynamically.

cmastalli replied:
Let's define a global variable for this, and called it `jMc_`.
This avoids to allocate data dynamically.

cmastalli commented on include/crocoddyl/multibody/contacts/contact-1d.hxx:
```diff
@@ -121,4 +145,24 @@ void ContactModel1DTpl<Scalar>::set_reference(const Scalar reference) {
+
+template <typename Scalar>
+void ContactModel1DTpl<Scalar>::set_mask(const Contact1DMaskType mask) {
+  mask_ = mask;
```
If you change to `LOCAL_WORLD_ALIGNED`, then define the identity matrix as pointed out above.

cmastalli replied:
If you change to `LOCAL_WORLD_ALIGNED`, then define the identity matrix as pointed out above.

cmastalli reviewed (changes requested):
Please see all the comments below.
A part of them, we also need to unittest the different mask cases.
To do so, you need to extend the contact factory.

skleff1994 commented:
Thanks for the review , I addressed the requested changes in [1fecfb4b](https://github.com/loco-3d/crocoddyl/pull/1054/commits/1fecfb4bc96b4ebf966e2a443ddaf87b1bc86cf1) . 

In addition to those , I implemented 
- force computation using mask and reference type for contact 1D
- specific unittest for 1D contact ( testing over masks and reference frames ) 

Currently, the unittest pass only in LOCAL and WORLD. There seem to be a mistake in calcDiff for LOCAL_WORLD_ALIGNED but I can't spot it..

skleff1994 commented:
So we wrote down the maths with @nmansard  and derived correct LOCAL_WORLD_ALIGNED derivatives for acceleration, I will fix the current implementation asap. 

Also @jcarpent  since the WORLD case doesn't make sense as you say I can remove it from the API. But then I think we should also clarify the Baumgarte terms: is there a reason why they are expressed in WORLD currently?

cmastalli commented:
I will close this issue due to a lack of activity. @skleff1994 -- feel free to re-open it anytime you wish.
